retrieval/targetmanager.go

```diff
@@ -14,139 +14,356 @@
-	Remove(t Target)
-	AddTargetsFromConfig(config config.Config)
+// A TargetProvider provides information about target groups. It maintains a set
+// of sources from which TargetGroups can originate. Whenever a target provider
+// detects a potential change it send the TargetGroup through its provided channel.
+//
+// The TargetProvider does not have to guarantee that an actual change happened.
+// It does guarantee that it sends the new TargetGroup whenever a change happens.
+// On startup it sends all TargetGroups it can see.
+type TargetProvider interface {
+	// Sources returns the source identifiers the provider is currently aware of.
+	Sources() []string
+	// Run hands a channel to the target provider through which it can send
+	// updated target groups. The channel must be closed by the target provider
+	// if no more updates will be sent.
+	Run(chan<- *config.TargetGroup)
+	// Stop terminates any potential computation of the target provider. The
+	// channel received on Run must be closed afterwards.
 	Stop()
-	Pools() map[string]*TargetPool // Returns a copy of the name -> TargetPool mapping.
 }
 
-type targetManager struct {
-	sync.Mutex     // Protects poolByJob.
+// TargetManager maintains a set of targets, starts and stops their scraping and
+// creates the new targets based on the target groups it receives from various
+// target providers.
+type TargetManager struct {
+	m              sync.Mutex
 	globalLabels   clientmodel.LabelSet
 	sampleAppender storage.SampleAppender
-	poolsByJob     map[string]*TargetPool
+	running        bool
+
+	// Targets by their source ID.
+	targets map[string][]Target
+	// Providers and configs by their job name.
+	// TODO(fabxc): turn this into map[*ScrapeConfig][]TargetProvider eventually.
+	providers map[string][]TargetProvider
+	configs   map[string]config.JobConfig
 }
 
-// NewTargetManager returns a newly initialized TargetManager ready to use.
-func NewTargetManager(sampleAppender storage.SampleAppender, globalLabels clientmodel.LabelSet) TargetManager {
-	return &targetManager{
+// NewTargetManager creates a new TargetManager based on the given config.
+func NewTargetManager(cfg config.Config, sampleAppender storage.SampleAppender) (*TargetManager, error) {
+	tm := &TargetManager{
 		sampleAppender: sampleAppender,
-		globalLabels:   globalLabels,
-		poolsByJob:     make(map[string]*TargetPool),
+		targets:        make(map[string][]Target),
+	}
+	if err := tm.updateConfig(cfg); err != nil {
+		return nil, err
 	}
+	return tm, nil
 }
 
-func (m *targetManager) targetPoolForJob(job config.JobConfig) *TargetPool {
-	targetPool, ok := m.poolsByJob[job.GetName()]
+// Run starts background processing to handle target updates.
+func (tm *TargetManager) Run() {
+	glog.Info("Starting target manager...")
 
-	if !ok {
-		var provider TargetProvider
-		if job.SdName != nil {
-			provider = NewSdTargetProvider(job, m.globalLabels)
-		}
+	sources := make(map[string]bool)
```

Comment:
Hmm, good question. In Prometheus, we got rid of `map[...]bool` in favor of `map[...]struct{}` everywhere a while ago (except for one last occurrence):

```
$ ag 'map\[.*\]bool' | grep -v Godeps | grep -v .build | wc -l
1
$ ag 'map\[.*\]struct{}' | grep -v Godeps | grep -v .build | wc -l
16
```

In a few cases, it could actually matter because of memory usage (an empty struct is 0 bytes, a boolean is 8 bytes). I guess we decided to just be consistent about it in the other places as well: when not actually using the boolean to store anything, use an empty struct to be clear about that (so it's almost more of a matter of clarity of intention than memory usage). I don't feel too strongly generally, but I think being consistent here has some value...